File: Linguagem Java/Questão12.java
//12. Sobrecarga de Operadores (Python) / Métodos de Conveniência (Java, Go) Em Python, sobrecarregue o operador + para somar dois objetos Produto baseados no preço. Em Java e Golang, crie métodos que permitam essa funcionalidade.


class Produto {
    private String nome;
    private double preco;

    public Produto(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    public Produto somar(Produto outro) {
        return new Produto("Soma de " + this.nome + " e " + outro.nome, this.preco + outro.preco);
    }

    @Override
    public String toString() {
        return String.format("%s: R$%.2f", nome, preco);
    }

    public static void main(String[] args) {
        Produto produto1 = new Produto("Nescau", 50.0);
        Produto produto2 = new Produto("Nesquick", 30.0);

        Produto produtoSoma = produto1.somar(produto2);
        System.out.println(produtoSoma);
    }
}

File: Linguagem Python/Questão12.py
#12. Sobrecarga de Operadores (Python) / Métodos de Conveniência (Java, Go) Em Python, sobrecarregue o operador + para somar dois objetos Produto baseados no preço. Em Java e Golang, crie métodos que permitam essa funcionalidade.


class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco

    def __add__(self, outro):
        if isinstance(outro, Produto):
            return Produto("Soma de " + self.nome + " e " + outro.nome, self.preco + outro.preco)
        return NotImplemented

    def __str__(self):
        return f"{self.nome}: R${self.preco:.2f}"


def main():
    produto1 = Produto("Nescau", 25.0)
    produto2 = Produto("Nesquick", 60.0)

    produto_soma = produto1 + produto2
    print(produto_soma)

if __name__ == "__main__":
    main()
